go/ql/src/experimental/CWE-522-DecompressionBombs/DecompressionBombs.ql

```diff
@@ -0,0 +1,68 @@
+/**
+ * @name Uncontrolled file decompression
+ * @description Uncontrolled data that flows into decompression library APIs without checking the compression rate is dangerous
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 7.8
+ * @precision high
+ * @id go/uncontrolled-file-decompression
+ * @tags security
+ *       experimental
+ *       external/cwe/cwe-409
+ */
+
+import go
+import semmle.go.dataflow.Properties
+import MultipartAndFormRemoteSource
+import experimental.frameworks.DecompressionBombs
+
+module DecompressionBombsConfig implements DataFlow::StateConfigSig {
+  class FlowState = DecompressionBombs::FlowState;
+
+  predicate isSource(DataFlow::Node source, FlowState state) {
+    source instanceof UntrustedFlowSource and
+    state = ""
+  }
+
+  predicate isSink(DataFlow::Node sink, FlowState state) {
+    sink instanceof DecompressionBombs::Sink and
+    state =
+      [
+        "ZstdNewReader", "XzNewReader", "GzipNewReader", "S2NewReader", "SnapyNewReader",
+        "ZlibNewReader", "FlateNewReader", "Bzip2NewReader", "ZipOpenReader", "ZipKlauspost"
+      ]
+  }
+
+  predicate isAdditionalFlowStep(DataFlow::Node fromNode, DataFlow::Node toNode) {
+    exists(DecompressionBombs::AdditionalTaintStep addStep |
+      addStep.isAdditionalFlowStep(fromNode, toNode)
+    )
+  }
+
+  predicate isAdditionalFlowStep(
+    DataFlow::Node fromNode, FlowState fromState, DataFlow::Node toNode, FlowState toState
+  ) {
+    exists(DecompressionBombs::AdditionalTaintStep addStep |
+      addStep.isAdditionalFlowStep(fromNode, fromState, toNode, toState)
+    )
+  }
+
+  predicate isBarrier(DataFlow::Node node) {
+    // here I want to the CopyN return value be compared with < or > but I can't reach the tainted result
+    exists(Function f | f.hasQualifiedName("io", "CopyN") |
+      node = f.getACall().getArgument(1) and
+      TaintTracking::localExprTaint(f.getACall().getResult(0).asExpr(),
+        // only >=, <=,>,<
+        any(RelationalComparisonExpr rce).getAnOperand())
+    )
```

Comment:
The first problem is that the two `f.getACall()` could be different.

The second problem is the assumption that `f.getACall().getResult(0)` is a `DataFlow::ExprNode` (and hence `asExpr()` has a value). It's actually a `DataFlow::InstructionNode` where the instruction is a `IR::extractTupleElement`.

After fixing that, taint still doesn't flow to an operand of a relational comparison. The third problem is that `totalBytes = totalBytes + result` doesn't make taint flow from `result` to `totalBytes`. I don't know the answer to this problem at the moment. When I next have time I will look into it.
```suggestion
    exists(Function f, DataFlow::CallNode cn |
      f.hasQualifiedName("io", "CopyN") and cn = f.getACall()
    |
      node = cn.getArgument(1) and
      TaintTracking::localTaint(cn.getResult(0),
        any(DataFlow::RelationalComparisonNode rcn).getAnOperand())
    )
```